File: Java/3sum.java
class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        
        Arrays.sort(nums);
        List<List<Integer>> res = new ArrayList<>(); 
        for(int i=0; i<nums.length; i++){
            if(i>0 && nums[i-1] == nums[i])
                continue;
            
            for(int j=i+1, k=nums.length-1; j<k; ){
                int sum = nums[i] + nums[j] + nums[k];
                if(sum == 0){
                    List<Integer> l = new ArrayList<>(); 
                    l.add(nums[i]);
                    l.add(nums[j]);
                    l.add(nums[k]);
                    res.add(l);
                    
                    int numj = nums[j];
                    while(j<k && nums[j]==numj)
                        j++;
                    
                    int numk = nums[k];
                    while(k>j && nums[k]==numk)
                        k--;
                    
                }else if(sum>0){
                    k--;
                }else{
                    j++;
                }
            }
        }
        
        return res;
    }
}

File: Python/3sum.py
class Solution:
    def threeSum(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        nums.sort()
        res = []
        for i in range(len(nums)):
            if i>0 and nums[i]==nums[i-1]:
                continue
                
            j=i+1
            k=len(nums)-1
            
            while j<k:
                vsum = nums[i] + nums[j] + nums[k]
                if vsum == 0:
                    res.append([nums[i], nums[j], nums[k]])
                    
                    numj = nums[j]
                    while j<k and numj == nums[j]:
                        j += 1
                        
                    numk = nums[k]
                    while k>j and numk == nums[k]:
                        k -= 1
                    
                elif vsum>0:
                    k -= 1
                else:
                    j += 1
                    
                    
        return res
